libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -0,0 +1,871 @@
+            return typedArray.getResourceId(
+                index,
+                defaultValue
+            ).also {
+                typedArray.recycle()
+            }
+        }
+
+        /**
+         * Checks if a layer with the given ID exists else returns a default layer ID
+         * @param layerId the layer ID to look for
+         * @param style the style containing the layers
+         *
+         * @return either the layer ID if found else a default layer ID
+         */
+        fun getBelowLayer(layerId: String?, style: Style): String =
+            style.layers.firstOrNull { it.id == layerId }?.id ?: LocationComponentConstants.SHADOW_LAYER
+
+        /**
+         * Generates a FeatureCollection and LineString based on the @param route.
+         * @param route the DirectionsRoute to used to derive the result
+         *
+         * @return a RouteFeatureData containing the original route and a FeatureCollection and
+         * LineString
+         */
+        fun generateFeatureCollection(route: DirectionsRoute): RouteFeatureData {
+            val routeGeometry = LineString.fromPolyline(
+                route.geometry() ?: "",
+                Constants.PRECISION_6
+            )
+            val routeFeature = Feature.fromGeometry(routeGeometry)
+
+            return RouteFeatureData(route, FeatureCollection.fromFeatures(listOf(routeFeature)), routeGeometry)
+        }
+
+        /**
+         * Creates an Expression that can be applied to the layer style changing the appearance of
+         * a route line, making the portion of the rout line behind the puck invisible.
+         *
+         * @param route the DirectionsRoute used for the Expression calculations
+         * @param routeLineString an optional LineString derived from the DirectionsRoute's geometry
+         * @param isPrimaryRoute indicates if the route used is the primary route
+         * @param distFractionToVanish indicates what portion of the route line should be invisible,
+         * usually based on a previous calculation. Traffic sections before this distance value won't
+         * be calculated.
+         * @param congestionColorProvider a function that provides the colors used for various
+         * traffic congestion values
+         *
+         * @return the Expression that can be used in a Layer's properties.
+         */
+        @JvmStatic
+        fun buildRouteLineExpression(
+            route: DirectionsRoute,
+            routeLineString: LineString?,
+            isPrimaryRoute: Boolean,
+            distFractionToVanish: Double,
+            congestionColorProvider: (String, Boolean) -> Int
+        ): Expression {
+            val vanishStep = Expression.stop(
+                distFractionToVanish,
+                Expression.rgba(0, 0, 0, 0)
+            )
+            val lineString: LineString = routeLineString ?: LineString.fromPolyline(route.geometry()!!, Constants.PRECISION_6)
```

Comment:
Could we ensure that `routeLineString` is always cached when route changes and provided by the caller instead of adding another logical branch here?

Reply:
The routeLineString is cached the MapRouteLine class. I thought it might be too restrictive to require it since the function can generate it but now that you point it out and I think about it more the caller can just as easily generate it an pass it into the function so I'll make it a non-nullable field.